mllib/src/main/scala/org/apache/spark/mllib/clustering/AffinityPropagation.scala

```diff
@@ -0,0 +1,347 @@
+   */
+  def this() = this(maxIterations = 100, lambda = 0.5, normalization = false)
+
+  /**
+   * Set maximum number of iterations of the messaging iteration loop
+   */
+  def setMaxIterations(maxIterations: Int): this.type = {
+    this.maxIterations = maxIterations
+    this
+  }
+ 
+  /**
+   * Get maximum number of iterations of the messaging iteration loop
+   */
+  def getMaxIterations(): Int = {
+    this.maxIterations
+  }
+ 
+  /**
+   * Set lambda of the messaging iteration loop
+   */
+  def setLambda(lambda: Double): this.type = {
+    this.lambda = lambda
+    this
+  }
+ 
+  /**
+   * Get lambda of the messaging iteration loop
+   */
+  def getLambda(): Double = {
+    this.lambda
+  }
+ 
+  /**
+   * Set whether to do normalization or not
+   */
+  def setNormalization(normalization: Boolean): this.type = {
+    this.normalization = normalization
+    this
+  }
+ 
+  /**
+   * Get whether to do normalization or not
+   */
+  def getNormalization(): Boolean = {
+    this.normalization
+  }
+ 
+  /**
+   * Run the AP algorithm.
+   *
+   * @param similarities an RDD of (i, j, s,,ij,,) tuples representing the similarity matrix, which
+   *                     is the matrix S in the AP paper. The similarity s,,ij,, is set to
+   *                     real-valued similarities. This is not required to be a symmetric matrix 
+   *                     and hence s,,ij,, can be not equal to s,,ji,,. Tuples with i = j are
+   *                     referred to as "preferences" in the AP paper. The data points with larger
+   *                     values of s,,ii,, are more likely to be chosen as exemplars.
+   *
+   * @param symmetric the given similarity matrix is symmetric or not. Default value: true
+   * @return a [[AffinityPropagationModel]] that contains the clustering result
+   */
+  def run(similarities: RDD[(Long, Long, Double)], symmetric: Boolean = true)
+    : AffinityPropagationModel = {
+    val s = constructGraph(similarities, normalization, symmetric)
+    ap(s)
+  }
+
+  /**
+   * Runs the AP algorithm.
+   *
+   * @param s The (normalized) similarity matrix, which is the matrix S in the AP paper with vertex
+   *          similarities and the initial availabilities and responsibilities as its edge
+   *          properties.
+   */
+  private def ap(s: Graph[Seq[Double], Seq[Double]]): AffinityPropagationModel = {
+    val g = apIter(s, maxIterations, lambda)
+    chooseExemplars(g)
+  }
+}
+
+private[clustering] object AffinityPropagation extends Logging {
+  /**
+   * Construct the similarity matrix (S) and do normalization if needed.
+   * Returns the (normalized) similarity matrix (S).
+   */
+  def constructGraph(similarities: RDD[(Long, Long, Double)], normalize: Boolean,
+    symmetric: Boolean):
+    Graph[Seq[Double], Seq[Double]] = {
+    val edges = similarities.flatMap { case (i, j, s) =>
+      if (symmetric && i != j) {
+        Seq(Edge(i, j, Seq(s, 0.0, 0.0)), Edge(j, i, Seq(s, 0.0, 0.0)))
```

Comment:
Using `Seq(s, 0.0, 0.0)` may have performance issues. If the tuples are used frequently, we can define a case class for it. It also makes the code easier to read. It is definitely not clear to me what the second and the third values are.